scripts/migration/courses-populate.ts

```diff
@@ -0,0 +1,172 @@
+/* eslint-disable no-console */
+/* eslint-disable no-await-in-loop */
+/**
+ * This script retrieves data from all courses from all available
+ * class rosters and pushes them to the 'courses' collection in
+ * Firebase.
+ *
+ * This script also creates a mapping from courses to the rosters
+ * that the course is available in.
+ *
+ *
+ * This script also creates a mapping from crseId to its
+ * catalogNbr.
+ *
+ *  Motivation for this script as well as the schema
+ * for the 'courses' collection can be found here:
+ * https://tinyurl.com/5n86j4p7
+ */
+
+import fetch from 'node-fetch';
+import { FieldValue } from 'firebase-admin/firestore';
+import {
+  coursesCollection,
+  availableRostersForCourseCollection,
+  crseIdToCatalogNbrCollection,
+} from '../firebase-config';
+import { CourseFullDetail } from '../../src/requirements/types';
+
+/** A helper function to generate a wait promise. Used for cooldown to limit API usage. */
+const wait = (time: number) =>
+  new Promise<void>(resolve => {
+    setTimeout(() => resolve(), time);
+  });
+
+/* Retrieves the rosters available on the class roster API */
+const retrieveAvailableRosters = async () => {
+  const rosters: readonly string[] = await fetch(
+    'https://classes.cornell.edu/api/2.0/config/rosters.json'
+  )
+    .then(res => res.json())
+    .then(jsonRes => jsonRes.data.rosters)
+    .then(jsonRosters => jsonRosters.map(jsonRoster => jsonRoster.slug));
+
+  return rosters;
+};
+
+/* Retrieves the subjects across for a roster and creates a list of {roster: string, subject: string} objects */
+const retrieveAvailableSubjects = (
+  roster: string
+): Promise<readonly { roster: string; subject: string }[]> =>
+  fetch(`https://classes.cornell.edu/api/2.0/config/subjects.json?roster=${roster}`)
+    .then(res => res.json())
+    .then(jsonRes => jsonRes.data.subjects)
+    .then(jsonSubjects =>
+      jsonSubjects.map(jsonSubject => ({ roster, subject: jsonSubject.value }))
+    );
```

Comment:
```suggestion
const retrieveAvailableSubjects = async (
  roster: string
): Promise<readonly { roster: string; subject: string }[]> => {
  const res = await fetch(`https://classes.cornell.edu/api/2.0/config/subjects.json?roster=${roster}`);
  return res.json().data.subjects.map(({ value }) => ({ roster, subject: value }))
}
```
Sorry to make multiple suggestions, but is there a reason something like this wouldn't work? I don't see the need for a bunch of chained `then`s. If so, can you also remove the other unneeded chained `then`s?

Reply:
Yeah I can do that, it is probably better to use `async` `await` syntax.